Cascading delete deployment

soltysh opened the pull request:
This implements reaper for deployments. I've had to modify `ObjectReaction` to get subresource from request, when specified and exists (1st commit) for update scale to work properly. Though this solution might not be perfect, need to test it more extensively. 
But most importantly current implementation of `GetOldRCs` does not return any RC if the pods where removed already, which unfortunately means that old RC don't get removed. 

@kargakis @nikhiljindal ideas?

Fixes #18078

soltysh committed:
Adding reaper for deployments

kargakis commented on pkg/kubectl/stop.go:
```diff
@@ -272,6 +280,54 @@ func (reaper *JobReaper) Stop(namespace, name string, timeout time.Duration, gra
+func (reaper *DeploymentReaper) Stop(namespace, name string, timeout time.Duration, gracePeriod *api.DeleteOptions) error {
+	deployments := reaper.Extensions().Deployments(namespace)
+	rcs := reaper.ReplicationControllers(namespace)
+	scaler, err := ScalerFor(extensions.Kind("Deployment"), *reaper)
```
I don't think you need to use a scaler for deployments here. Just get all the rcs of the deployment and reap them. Delete the deployment first so that the deployment controller won't recreate it in the meantime. See https://github.com/openshift/origin/blob/master/pkg/deploy/reaper/reaper.go for our existing deployment reaper.

kargakis commented on pkg/kubectl/stop.go:
```diff
@@ -272,6 +280,54 @@ func (reaper *JobReaper) Stop(namespace, name string, timeout time.Duration, gra
+		timeout = Timeout + time.Duration(10*deployment.Spec.Replicas)*time.Second
+	}
+
+	// TODO: handle overlapping deployments
```
There shouldn't be any overlapping deployments, remove this TODO

kargakis commented:
@bgrant0607 had an opinion on reaping deployments (server-side vs client-side)

Also cc: @ironcladlou @janetkuo

bgrant0607 commented:
We can start with client-side if that's easiest, but at some point I want to implement the existence dependency mechanism mentioned in #1535, #3585, and elsewhere, and have a GC controller clean up.

kargakis commented on hack/test-cmd.sh:
```diff
@@ -891,8 +891,6 @@ __EOF__
-  kubectl delete rc --all "${kube_flags[@]}"
 
   ### Expose replication controller as service
   kubectl create -f examples/guestbook/frontend-controller.yaml "${kube_flags[@]}"
```
Remove this create as well and make the following pre-condition 2 from 3.

soltysh replied:
Not needed, since we're creating 3 in the above create command and expect them to exists.

kargakis commented on pkg/client/unversioned/testclient/fixture.go:
```diff
@@ -60,6 +60,9 @@ func ObjectReaction(o ObjectRetriever, mapper meta.RESTMapper) ReactionFunc {
 	return func(action Action) (bool, runtime.Object, error) {
 		gvk, err := mapper.KindFor(action.GetResource())
+		if len(action.GetSubresource()) > 0 {
+			gvk, _ = mapper.KindFor(action.GetSubresource())
```
Why do you need a subresource?

kargakis replied:
ah because of deployment/scale. You don't need to scale the deployment down. What you need is:

1) GET deployment/name
2) GET rcs from deployment/name
3) STOP deployment/name
4) STOP rcs

kargakis commented on pkg/kubectl/stop.go:
```diff
@@ -95,6 +99,10 @@ type JobReaper struct {
 	pollInterval, timeout time.Duration
 }
+type DeploymentReaper struct {
+	client.Interface
```
use an extensions client

kargakis replied:
hm, actually you need to delete rcs too. nvmd

kargakis commented:
Fixes https://github.com/kubernetes/kubernetes/issues/18078

soltysh commented:
Waiting for https://github.com/kubernetes/kubernetes/pull/18169

kargakis commented:
Side note: @soltysh clean-up this [TODO](https://github.com/kubernetes/kubernetes/blob/2efc738d5ba221a93c05e2a9a2c44738613e2e14/hack/test-cmd.sh#L941) when you'll touch this PR again.

kargakis commented:
> Waiting for #18169

FIY, this PR is now in master

soltysh commented:
Yup, I've seen it, I'll be working on it next week.

bgrant0607 commented:
cc @janetkuo

bgrant0607 commented:
cc @madhusudancs

bgrant0607 commented:
This has to get done next week.

Why can't Deployment's Stop just call Stop on all of its RCs?

bgrant0607 commented:
Nevermind, that wouldn't work. Why can't it just resize the Deployment to 0?

bgrant0607 commented:
Actually, once #19590 goes it, perhaps it could set RevisionHistoryLimit to 0 and then resize to 0.

kargakis commented:
> Actually, once #19590 goes in, perhaps it could set RevisionHistoryLimit to 0 and then resize to 0.

Good idea. I didn't realize we almost implemented reaping in the controller.

soltysh commented:
Working on it now..

soltysh committed:
Updated e2e description in stopping jobs

soltysh commented:
Updated and addressed comments. 
@kargakis @nikhiljindal ptal

kargakis commented on pkg/kubectl/stop.go:
```diff
@@ -276,6 +283,69 @@ func (reaper *JobReaper) Stop(namespace, name string, timeout time.Duration, gra
+	// clear deployment history
+	zero := 0
+	deployment.Spec.RevisionHistoryLimit = &zero
+	deployment, err = deployments.Update(deployment)
```
Why not update both d.spec.replicas and d.spec.revisionHistoryLimit? Then you won't need to scale down since you already poll d.status.replicas

soltysh replied:
Done.

kargakis commented on pkg/kubectl/stop.go:
```diff
@@ -276,6 +283,69 @@ func (reaper *JobReaper) Stop(namespace, name string, timeout time.Duration, gra
+		return utilerrors.NewAggregate(errList)
+	}
+
+	if err := deployments.Delete(name, gracePeriod); err != nil {
```
return deployments.Delete(name, gracePeriod)

soltysh replied:
Done, also for other methods.

kargakis commented on pkg/kubectl/stop.go:
```diff
@@ -276,6 +283,69 @@ func (reaper *JobReaper) Stop(namespace, name string, timeout time.Duration, gra
+	}
+	errList := []error{}
+	for _, rc := range rcList.Items {
+		if err := replicationControllers.Delete(rc.Name); err != nil {
```
ignore notFound errors

soltysh replied:
Done.

kargakis commented on pkg/kubectl/stop.go:
```diff
@@ -276,6 +283,69 @@ func (reaper *JobReaper) Stop(namespace, name string, timeout time.Duration, gra
+	selector := labels.SelectorFromSet(deployment.Spec.Selector)
+	options := api.ListOptions{LabelSelector: selector}
+	rcList, err := replicationControllers.List(options)
+	if err != nil {
```
if we get a notFound error from a list, you should ignore it

soltysh replied:
You'll get an empty list, not an error here.

soltysh commented:
@kargakis comments addressed

kargakis commented on hack/test-cmd.sh:
```diff
@@ -945,8 +945,6 @@ __EOF__
   kube::test::get_object_assert 'deployment nginx-deployment' "{{$deployment_replicas}}" '1'
   # Clean-up
   kubectl delete deployment/nginx-deployment "${kube_flags[@]}"
-  # TODO: Remove once deployment reaping is implemented
```
There is one more TODO you need to remove at L461

soltysh replied:
Done.

kargakis commented:
LGTM

soltysh commented:
@k8s-bot unit test this

soltysh commented:
I had to drop the commit modifying `ObjectReaction` to get subresource from request, it was causing tests to fail and it's not needed anymore.

ironcladlou commented:
I'd like to point out that embedding a specific workflow for stopping in the client is a potential source of technical debt:

1. The client behavior embedded in the binary is immutable for some (potentially large) set of deployed clients which may not be upgraded any time soon.
2. The specific sequence/composition of client operations we're cooking up to "stop" a deployment might not produce the same behavior later, causing unknowable amounts of future work to maintain compatible behavior with old clients absent any way of identifying a sequence of apparently disjoint API calls as part of an overall atomic workflow operation (e.g. "stop"). This bit us hard in OpenShift with very similar workflows embedded in clients with which we eventually had to maintain **behavioral** compatibility, even though API compatibility was maintained.

I don't know how concerning this is for the k8s CLI in this context or what the compatibility goals are, but I felt it worth mentioning the dangers.

soltysh commented:
With the test failures I dug deeper and I had to clean-up the test-cmd.sh tests a bit with necessary cleanups where missing. But most importantly I had to extend the stop deployment process with pausing deployment so that it don't create new RCs while others are being removed. Here's the current flow:
1. set RevisionHistoryLimit and Replicas to 0
2. wait for deployment.Status.Replicas drops to 0
3. pause the deployment
4. remove related RCs
5. remove deplyment itself.
@kargakis @bgrant0607 ptal

kargakis commented:
Cleanup policy should work for paused deployments. Thanks for uncovering this.

bgrant0607 commented:
cc @janetkuo

bgrant0607 commented on pkg/kubectl/stop.go:
```diff
@@ -270,10 +271,70 @@ func (reaper *JobReaper) Stop(namespace, name string, timeout time.Duration, gra
+	zero := 0
+	deployment.Spec.RevisionHistoryLimit = &zero
+	deployment.Spec.Replicas = 0
+	deployment, err = deployments.Update(deployment)
```
This is likely to get a resourceVersion conflict due to status updates. How about patch instead?

soltysh replied:
I don't have Patch operation available at this point, since I'm interacting with client.Interface at this level.

bgrant0607 replied:
Ok

soltysh replied:
I'll add TODO with link the the issue you've created.

bgrant0607 commented on pkg/kubectl/stop.go:
```diff
@@ -270,10 +271,70 @@ func (reaper *JobReaper) Stop(namespace, name string, timeout time.Duration, gra
+		if err != nil {
+			return false, nil
+		}
+		return curr.Status.Replicas == 0, nil
```
This is somewhat racy in the case of deletion right after creation, but we can address that later.

soltysh replied:
I've discussed it with @smarterclayton, he suggested adding checking NotFound error and UID - added.

kargakis replied:
FYI, RCs use generation to workaround this. Unfortunately we still don't have generation for Deployments.

soltysh replied:
Will add yet another TODO ;)

bgrant0607 commented on pkg/kubectl/stop.go:
```diff
@@ -270,10 +271,70 @@ func (reaper *JobReaper) Stop(namespace, name string, timeout time.Duration, gra
+		return err
+	}
+
+	// remove remaining RCs
```
Shouldn't we be able to just wait for the controller to do this, since we set the revision history limit to 0?

bgrant0607 replied:
Ah, this is because we paused the Deployment. Maybe we should clear "Paused" instead?

soltysh replied:
There still be that one last active RC with replicas set to 0, which we need to remove.

bgrant0607 replied:
Ok, but why are we setting Paused to true? We want the controller to sync the underlying RCs to the updated replica count and history limit.

janetkuo replied:
IIUC, we set Paused=true to stop the deployment from creating new RC. Perhaps in deployment controller, we shouldn't create new RC when deployment.Spec.Replicas=0 (currently it creates a rc with 0 replicas)?

kargakis replied:
> Perhaps in deployment controller, we shouldn't create new RC when deployment.Spec.Replicas=0 (currently it creates a rc with 0 replicas)?

No, this is intentional. I am going to update the behavior of the controller during paused to allow running the cleanup policy among other things. Then we can simplify the reaper to set d.spec.paused=true, d.spec.revisionHistoryLimit=0, d.spec.replicas=0 in one update, then have it poll status.replicas, delete latest rc, delete deployment.

soltysh replied:
I'll add TODO to fix this once your feature will be in.

bgrant0607 replied:
This comment still says "RCs"

bgrant0607 commented:
Good points @ironcladlou. Since this API will be beta, I'm not hugely concerned. I would like to implement cascading delete server-side before this goes to v1, though.

janetkuo commented on hack/test-cmd.sh:
```diff
@@ -668,9 +668,9 @@ runTests() {
-  # Clean up 
+  # Clean up
   kubectl delete deployment nginx "${kube_flags[@]}"
   kubectl delete rc -l deployment.kubernetes.io/podTemplateHash "${kube_flags[@]}"
```
Remove this line too.

janetkuo commented on hack/test-cmd.sh:
```diff
@@ -1064,7 +1063,7 @@ __EOF__
   kubectl delete rc frontend "${kube_flags[@]}"
 
-  ### Auto scale deployment 
+  ### Auto scale deployment
```
Remove L1078   `kubectl delete rc -l deployment.kubernetes.io/podTemplateHash "${kube_flags[@]}"`

janetkuo commented on test/e2e/deployment.go:
```diff
@@ -149,6 +152,23 @@ func checkDeploymentRevision(c *clientset.Clientset, ns, deploymentName, revisio
+	deployment, err := c.Extensions().Deployments(ns).Get(deploymentName)
+	Expect(err).NotTo(HaveOccurred())
+
+	Logf("stopping deployment %s", deploymentName)
```
Suggest that we use the term "delete" instead of "stop" since `kubectl stop` is deprecated and superseded by `delete`.

kargakis replied:
+1

janetkuo commented on test/e2e/deployment.go:
```diff
@@ -149,6 +152,23 @@ func checkDeploymentRevision(c *clientset.Clientset, ns, deploymentName, revisio
+	Logf("ensuring deployment %s was stopped", deploymentName)
+	_, err = c.Extensions().Deployments(ns).Get(deployment.Name)
+	Expect(err).To(HaveOccurred())
+	Expect(errors.IsNotFound(err)).To(BeTrue())
```
Ensure the rcs and pods are gone, too?

soltysh replied:
I'll add just checking rcs, pods have grace period, which I'd have to wait, not sure if it's worth.

soltysh replied:
Actually after switching to rc reaper instead of plain delete it's working as expected, adding pods as well.

janetkuo commented on test/e2e/deployment.go:
```diff
@@ -149,6 +152,23 @@ func checkDeploymentRevision(c *clientset.Clientset, ns, deploymentName, revisio
 	return deployment, newRC
 }
 
+func stopDeployment(c *clientset.Clientset, oldC client.Interface, ns, deploymentName string) {
```
oldC will be removed soon in #20452. f.Client will return *clientset.Clientset. cc @caesarxuchao

caesarxuchao replied:
Yes, I'll remove the oldC in O(days).

caesarxuchao replied:
It will probably take longer time. I'll do the rebase.

kargakis replied:
@soltysh +1 for this helper

soltysh commented:
Addressed all comments.

soltysh commented:
@bgrant0607 @janetkuo @kargakis ptal

soltysh commented:
grrrr... forgot to update unittests - fixed.

bgrant0607 commented on pkg/kubectl/stop.go:
```diff
@@ -270,10 +271,84 @@ func (reaper *JobReaper) Stop(namespace, name string, timeout time.Duration, gra
+		return err
+	}
+
+	// TODO: when deployment will allow running cleanup policy when paused
```
This is hard to read. Please add some punctuation between sentences.

bgrant0607 replied:
And fix line breaks

soltysh replied:
Done.

bgrant0607 commented on pkg/kubectl/stop.go:
```diff
@@ -270,10 +271,84 @@ func (reaper *JobReaper) Stop(namespace, name string, timeout time.Duration, gra
+	// remove remaining RCs
+	selector := labels.SelectorFromSet(deployment.Spec.Selector)
+	options := api.ListOptions{LabelSelector: selector}
+	rcList, err := replicationControllers.List(options)
```
@kargakis @janetkuo How does Deployment find the RCs that it controls? Does it assume that they are labeled with the same labels as the pod template?

kargakis replied:
yes

janetkuo replied:
Deployment finds the RCs with labels that match the deployment's selector. See https://github.com/kubernetes/kubernetes/blob/master/pkg/util/deployment/deployment.go#L65

@soltysh Could you make this a function to make sure deployment is using the same way to retrieve the RCs it targets, when getOldRCs, getNewRC and stop deployment?

soltysh commented:
@k8s-bot test this github issue: #18078

bgrant0607 commented:
My questions were answered. I'll let @kargakis and @janetkuo give the final LGTM.

kargakis commented:
LGTM

@soltysh thanks for working on this! Will buy you a beer this weekend:)

soltysh commented:
Rebased.

janetkuo commented:
Code looks good. However, some e2e tests failed locally because of redundant `defer` functions that clean up RCs. Would you fix that and address https://github.com/kubernetes/kubernetes/pull/18077#discussion_r51794515 too? I'll apply LGTM tag once it's done. Thanks much!

soltysh commented:
@janetkuo not sure which failures are you talking about, everything seems to be working ok. Regarding the comment you pointed out, I was originally using `GetOldRCsFromLists` but in the mean time `clientsets` where introduced, I had to I fall back to manually getting the required RCs the same way the aforementioned method does. I'm not sure it's worth redoing it with current changes to clientset going on.

soltysh commented:
Rebased.

soltysh commented:
@k8s-bot test this github issue: #18078

bgrant0607 commented:
@soltysh The next time you rebase this, let us know and we'll try to help get it merged.

bgrant0607 commented:
That is, @ notify me.

janetkuo commented:
@soltysh I should clarify. All deployment e2e tests are removed from building test (see #19667 and https://github.com/kubernetes/kubernetes/blob/master/docs/devel/e2e-tests.md#kinds-of-tests) so that they won't block merge queue. So you won't see those test failures here. I tried downloading the code and tested them locally and found several failures. 

Please avoid merging this before making sure all deployment e2e tests passed. cc @bgrant0607 

The failure that I mentioned earlier is that the adopted RCs are deleted with `defer`, but we used another `defer` to clean up deployments and then there'll be RC not found error while we try to delete it. 

Here are some of the failures I saw from running all deployment e2e tests:
```
RecreateDeployment should delete old pods and create new ones [It]
src/k8s.io/kubernetes/test/e2e/deployment.go:49
  Expected error:
      <*errors.errorString | 0xc820182d50>: {
          s: "timed out waiting for the condition",
      }
      timed out waiting for the condition
  not to have occurred
/usr/local/google/home//go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/deployment.go:375
```
```
  RollingUpdateDeployment should delete old pods and create new ones [It]
  /usr/local/google/home//go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/deployment.go:43

  Expected error:
      <*errors.errorString | 0xc820182d50>: {
          s: "timed out waiting for the condition",
      }
      timed out waiting for the condition
  not to have occurred
  /usr/local/google/home//go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/deployment.go:247
```
```
  deployment should delete old rcs [It]
  /usr/local/google/home//go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/deployment.go:52

  Expected error:
      <*errors.errorString | 0xc820182d50>: {
          s: "timed out waiting for the condition",
      }
      timed out waiting for the condition
  not to have occurred
 /usr/local/google/home//go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/deployment.go:440
```
```
  paused deployment should be ignored by the controller [It]
  /usr/local/google/home//go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/deployment.go:58

  Test Panicked
  runtime error: invalid memory address or nil pointer dereference
  /usr/local/go/src/runtime/panic.go:423

  Full Stack Trace
  /usr/local/go/src/runtime/panic.go:423 (0x42c759)
        gopanic: reflectcall(nil, unsafe.Pointer(d.fn), deferArgs(d), uint32(d.siz), uint32(d.siz))
  /usr/local/go/src/runtime/panic.go:42 (0x42ac89)
        panicmem: panic(memoryError)
  /usr/local/go/src/runtime/sigpanic_unix.go:24 (0x441bca)
        sigpanic: panicmem()
  /usr/local/google/home//go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/deployment.go:580 (0x4aeed6)
        testPausedDeployment: // Create a deployment to create nginx pods.
  /usr/local/google/home//go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/deployment.go:57 (0x5c60f0)
        glob.func13.7: })
  /usr/local/google/home//go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/e2e.go:239 (0x4bd800)
        RunE2ETests: ginkgo.RunSpecsWithDefaultAndCustomReporters(t, "Kubernetes e2e suite", r)
```
```
/usr/local/google/home//go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/deployment.go:65
  deployment should create new pods [It]
  /usr/local/google/home//go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/deployment.go:40

  Test Panicked
  runtime error: invalid memory address or nil pointer dereference
  /usr/local/go/src/runtime/panic.go:423

  Full Stack Trace
  /usr/local/go/src/runtime/panic.go:423 (0x42c759)
        gopanic: reflectcall(nil, unsafe.Pointer(d.fn), deferArgs(d), uint32(d.siz), uint32(d.siz))
  /usr/local/go/src/runtime/panic.go:42 (0x42ac89)
        panicmem: panic(memoryError)
  /usr/local/go/src/runtime/sigpanic_unix.go:24 (0x441bca)
        sigpanic: panicmem()
  /usr/local/google/home//go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/deployment.go:175 (0x58d407)
        testNewDeployment.func1: Expect(rcs.Items).Should(HaveLen(0))
  /usr/local/go/src/runtime/asm_amd64.s:438 (0x45e25e)
        call64: CALLFN(·call64, 64)
  /usr/local/go/src/runtime/panic.go:423 (0x42c759)
        gopanic: reflectcall(nil, unsafe.Pointer(d.fn), deferArgs(d), uint32(d.siz), uint32(d.siz))
  /usr/local/go/src/runtime/panic.go:42 (0x42ac89)
        panicmem: panic(memoryError)
  /usr/local/go/src/runtime/sigpanic_unix.go:24 (0x441bca)
        sigpanic: panicmem()
  /usr/local/google/home//go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/deployment.go:136 (0x49f2ee)
        checkDeploymentRevision: // Check revision of the new RC of this deployment
```
```
  deployment should support rollback when there's RC with no revision [It]
  /usr/local/google/home//go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/deployment.go:64

  Test Panicked
  runtime error: invalid memory address or nil pointer dereference
  /usr/local/go/src/runtime/panic.go:423

  Full Stack Trace
  /usr/local/go/src/runtime/panic.go:423 (0x42c759)
        gopanic: reflectcall(nil, unsafe.Pointer(d.fn), deferArgs(d), uint32(d.siz), uint32(d.siz))
  /usr/local/go/src/runtime/panic.go:42 (0x42ac89)
        panicmem: panic(memoryError)
  /usr/local/go/src/runtime/sigpanic_unix.go:24 (0x441bca)
        sigpanic: panicmem()
  /usr/local/google/home//go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/deployment.go:738 (0x593ad1)
        testRollbackDeploymentRCNoRevision.func2: _, err = c.Extensions().Deployments(ns).Update(d)
```
```
  deployment should support rollover [It]
  /usr/local/google/home//go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/deployment.go:55

  Test Panicked
  runtime error: invalid memory address or nil pointer dereference
  /usr/local/go/src/runtime/panic.go:423

  Full Stack Trace
  /usr/local/go/src/runtime/panic.go:423 (0x42c759)
        gopanic: reflectcall(nil, unsafe.Pointer(d.fn), deferArgs(d), uint32(d.siz), uint32(d.siz))
  /usr/local/go/src/runtime/panic.go:42 (0x42ac89)
        panicmem: panic(memoryError)
  /usr/local/go/src/runtime/sigpanic_unix.go:24 (0x441bca)
        sigpanic: panicmem()
  /usr/local/google/home//go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/deployment.go:496 (0x5913d7)
```

janetkuo commented:
The easies way to run e2e tests locally is:

1. start a local cluster: `RUNTIME_CONFIG=extensions/v1beta1/deployments=true hack/local-up-cluster.sh`
2. `make all WHAT=test/e2e/e2e.test`
3. `e2e.test --host="127.0.0.1:8080" --provider="local" --ginkgo.focus="eployment should" --kubeconfig="$HOME/.kube/config-local" --repo-root="<YOUR_K8S_REPO_ROOT>"`

or if you used a GCE cluster instead, run:
`KUBE_CONFIG_FILE=config-default.sh hack/ginkgo-e2e.sh --ginkgo.focus="eployment should"`

soltysh commented:
@janetkuo thank you, I haven't seen those failure on my end. Will double check while rebasing.

soltysh commented:
@janetkuo rebased and fixed, sorry for previous problems, for some reason I was missing those tests. I've fixed it by removing those deferred calls, I've also added wait.Poll for waiting for pods to be actually removed in stopDeployment, because I was hitting very often a problem where a pod would be left in terminating state.

janetkuo commented:
@soltysh Thanks! The failures went away but I think it's flaky because sometimes it didn't remove pods:
```
paused deployment should be ignored by the controller [It]
/usr/local/google/home//go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/deployment.go:62
  Feb  6 07:38:08.848: Failed to remove deployment nginx pods!
/usr/local/google/home//go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/deployment.go:188
```

soltysh commented:
@janetkuo I've added code that waits for the pods to be terminated by the kubelet, all I can do now is to increase the timeout. Other option, I'm not in favor of, is removing pod checks.

bgrant0607 commented:
Set terminationGracePeriodSeconds to 0

soltysh commented:
@janetkuo I've set terminationGracePeriodSeconds to 0 to all pods created in deployments tests per @bgrant0607 advice (Thanks!). This allowed me to drop the waiting logic in stopDeployment. Hopefully that's the final solution.

soltysh commented:
@k8s-bot test this github issue: #18078

janetkuo commented:
@soltysh I tried testing it locally and one of the deployment e2e tests is failing pretty often (~50% of the time): `paused deployment should be ignored by the controller`. There's still one pod left after deployment and rcs are gone. 

The pod failed to get deleted looks like this:
```
          {
              TypeMeta: {Kind: "", APIVersion: ""},
              ObjectMeta: {
                  Name: "nginx-7yro4-8kcoo",
                  GenerateName: "nginx-7yro4-",
                  Namespace: "e2e-tests-deployment-xh25b",
                  SelfLink: "/api/v1/namespaces/e2e-tests-deployment-xh25b/pods/nginx-7yro4-8kcoo",
                  UID: "a833b04f-ceb3-11e5-adf7-64510658e388",
                  ResourceVersion: "6618",
                  Generation: 0,
...
                  DeletionTimestamp: nil,
                  DeletionGracePeriodSeconds: nil,
                  Labels: {
                      "name": "nginx",
                      "pod-template-hash": "1900829565",
                  },
                  Annotations: {
                      "kubernetes.io/created-by": "{\"kind\":\"SerializedReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"ReplicationController\",\"namespace\":\"e2e-tests-deployment-xh25b\",\"name\":\"nginx-7yro4\",\"uid\":\"a82c0691-ceb3-11e5-adf7-64510658e388\",\"apiVersion\":\"v1\",\"resourceVersion\":\"6610\"}}\n",
                  },
              },
...
```

After looking closer to the code in `testPausedDeployment`, I'm thinking that maybe `deployment.ResourceVersion = ""` is the cause? Why do we need this line there? cc @kargakis

kargakis commented:
@janetkuo that's for avoiding update conflicts

soltysh commented:
@janetkuo to address your problems I've reverted the logic responsible for waiting for pods to be gone in stopDeployments. I've rerun the entire deployments test suite a couple of times and didn't hit any problems. ptal

soltysh commented:
Rebased to address that last failure.

kargakis commented:
still lgtm

janetkuo commented:
@soltysh `paused deployment should be ignored by the controller` is still failing ~30% times on my machine. I think I figured out why. `testPausedDeployment` sets `deployment.Spec.Paused = true`, and this affected clean up process in `DeploymentReaperStop`. 

You need to "unpause" when setting RevisionHistoryLimit and Replicas in `Stop`. 
```
	zero := 0
	deployment.Spec.RevisionHistoryLimit = &zero
	deployment.Spec.Replicas = 0
	deployment.Spec.Paused = false  // We need to un-pause!
	deployment, err = deployments.Update(deployment)
```

Adding this change made the test passed 100% on my machine.

janetkuo commented:
@madhusudancs this is related to your PR #19840. This deletes deployments cascadingly by deleting its RCs/Pods.

janetkuo commented:
This will need a rebase to replace replication controllers with replica sets if #19840 goes in first.

soltysh commented:
@janetkuo rebased and addressed your comment. Additionally I've put the link in TODO to the policy cleanup for deployment issue. @bgrant0607 this should be finally ready. I apologize for the long-ish delay on moving this forward.

bgrant0607 commented:
@soltysh Sorry, but the rebase of Deployment onto ReplicaSet was merged yesterday. In addition to needing to update this to deal with ReplicaSets instead of ReplicationController, scale functionality is temporarily broken. If scale could be temporarily reimplemented to use a regular update instead of the scale subresource, we could make that work again.

How much time do you have to spend on this? Could you use help?

cc @madhusudancs @mqliang

soltysh commented:
Let me look into this in the morning. I have @kargakis at hand if needed.
I'll try to fix it tomorrow, I have some bandwidth.
On Feb 10, 2016 10:09 PM, "Brian Grant" <notifications@github.com> wrote:

> @soltysh <https://github.com/soltysh> Sorry, but the rebase of Deployment
> onto ReplicaSet was merged yesterday. In addition to needing to update this
> to deal with ReplicaSets instead of ReplicationController, scale
> functionality is temporarily broken. If scale could be temporarily
> reimplemented to use a regular update instead of the scale subresource, we
> could make that work again.
>
> How much time do you have to spend on this? Could you use help?
>
> cc @madhusudancs <https://github.com/madhusudancs> @mqliang
> <https://github.com/mqliang>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18077#issuecomment-182582776>
> .
>

madhusudancs commented:
@soltysh Also, let me know if you need some help. I am happy to help. (I did the ReplicationController -> ReplicaSet migration).

mqliang commented:
@soltysh I am also glad to help. Feel free to @madhusudancs or me if you need.

soltysh committed:
Scale deployments fall-back to regular deployment update

soltysh commented:
@bgrant0607 @madhusudancs @mqliang as suggested I've enabled the scale with full deployment update. Only slight changes where needed in a couple of places (see last commit). ptal where appropriate I've left @madhusudancs todo note.

bgrant0607 commented on pkg/kubectl/stop_test.go:
```diff
@@ -495,6 +496,133 @@ func TestJobStop(t *testing.T) {
+						Replicas: 0,
+					},
+				},
+				&extensions.Scale{ // UPDATE
```
The Scale subresource doesn't work currently. What is this testing?

bgrant0607 commented on pkg/kubectl/stop_test.go:
```diff
@@ -495,6 +496,133 @@ func TestJobStop(t *testing.T) {
+			Name: "Deployment with single replicaset",
+			Objs: []runtime.Object{
+				&deployment, // GET
+				&extensions.Scale{ // UPDATE
```
Scale again.

bgrant0607 commented:
I only had time to take a glance, and may not have time to get back to it for a few hours. I'd appreciate it if someone else did a minute thorough review before then.

bgrant0607 commented:
Tests failed

bgrant0607 commented:
cc @madhusudancs

soltysh commented:
@bgrant0607 I've commented the pieces using scale inside the fallback commit. The test failure was caused by moving deployment example to user-docs, I've rebased to address this problem. ptal again

soltysh commented:
@madhusudancs I've removed all the comments per your request, I've just left there TODO note we should rework full update to scale. ptal
/cc @bgrant0607

bgrant0607 commented:
Thanks, @soltysh. LGTM. We'll revisit it again as soon as /scale is implemented.

soltysh commented:
Thank you @bgrant0607 and all the other reviewers!
